pkg/cluster/loadbalancerprofile.go

```diff
@@ -263,8 +317,31 @@ func (m *manager) reconcileDesiredManagedIPs(ctx context.Context) ([]api.Resourc
 			break
 		}
 	}
+	return desiredIPAddresses
+}
 
-	return desiredIPAddresses, nil
+func (m *manager) createPublicIPAddresses(ctx context.Context, ipAddresses map[string]mgmtnetwork.PublicIPAddress, numToCreate int) error {
+	ch := make(chan createIPResult)
+	defer close(ch)
+	var errResults []string
+	// create additional IPs if needed
+	for i := 0; i < numToCreate; i++ {
+		go m.createPublicIPAddress(ctx, ch)
+	}
+
+	for i := 0; i < numToCreate; i++ {
+		result := <-ch
+		if result.err != nil {
+			errResults = append(errResults, fmt.Sprintf("creation of ip address %s failed with error: %s", *result.ip.Name, result.err.Error()))
+		} else {
+			ipAddresses[*result.ip.Name] = result.ip
```

Comment:
I have to admit that the fact that this function is modifying the ipAddresses map at the same time that it checks for the result of the ch channel is a bit unexpected/confusing for me. 

I think we could make "createPublicIPAddress" write into two different channels, the resultsChannels where we just write the ones that did not fail and an error channel that just receives and expects errors in there. In that way, we do not need to check with an if in "createPublicIPAddresses" if the operation succeded for the result of the channel as it is not easy to follow. We could simply iterate over the two channels, having a good separation of concerns. I think that would simplify the logic and would be more idiomatic and maintainable. I have worked on something that is working (happy the tests passed after some fight) and I would be happy to talk about it if you are interested in 🙂

Reply:
Update: we already discussed it privately. Nice chat.